### ISSUE TITLE
Small UI Bug

### ISSUE THREAD

**Issue by SentaiBrad** [sonic2kk/steamtinkerlaunch#638]
## System Information
- SteamTinkerLaunch version: Latest Git
- Distribution: SteamOS
- Installation Method: ProtonUp-Qt

## Issue Description
When STL launches a window that has a scroll bar, it displays this scrollbar on the top of any content in the window. Most windows have check boxes on the left, but there are some windows with a check box on the right, and the Scroll Bar covers this up, making them unable to be clicked or unclicked.

Of course, when I go to make the report I can not find the window that had the specific issue. However, here is a screenshot that still illustrates this.

![image](https://user-images.githubusercontent.com/26223818/197427741-08ab103d-e4e4-4a39-85f4-ba851e29a98e.png)

**Comment by sonic2kk** [comment]
This is only on Steam Deck and the wider scrollbar is intentional to make it easier to scroll. However, it was decreased in size in a recent version bump for the exact reason you outlined.

In your screenshot though it looks wider than it should be. Maybe SteamTinkerLaunch didn't update, which would be a concern... Are you using 20221024 ? It was updated before this, but if you were on this version then it definitely should have updated the scrollbar size...

**Comment by SentaiBrad** [comment]
> This is only on Steam Deck and the wider scrollbar is intentional to make it easier to scroll. However, it was decreased in size in a recent version bump for the exact reason you outlined.
> 
> In your screenshot though it looks wider than it should be. Maybe SteamTinkerLaunch didn't update, which would be a concern... Are you using 20221024 ? It was updated before this, but if you were on this version then it definitely should have updated the scrollbar size...

Yes, I'm on the most recent Git, for when I was testing the issue of the notifications.

To fully remedy this, would it be better to not draw the scroll bar over text or anything else?

**Comment by sonic2kk** [comment]
> To fully remedy this, would it be better to not draw the scroll bar over text or anything else?

I'm not totally sure, but I don't think this is possible with Yad, or most native windowing toolkits

**Comment by SentaiBrad** [comment]
> > To fully remedy this, would it be better to not draw the scroll bar over text or anything else?
> 
> I'm not totally sure, but I don't think this is possible with Yad, or most native windowing toolkits

I found the issues in question. The problem affects checkboxes less, more the + and - for some input fields (and maybe 1 or 2 checkboxes). It is also hard to highlight the numbers, and it's even worse in SteamOS GameMode because taping just hits the scroll wheel. No Ctrl+A, but it can be navigated by the on-screen keyboard. Clicking in the field, Ctrl + A in desktop mode, and typing in a new number does work.

**Comment by sonic2kk** [comment]
There is nothing really that can be done about things being hard to tap, that's kind of the problem with UI apps running on a handheld with a touchscreen. Yad, and by extension regular GTK windowing toolkits (and Qt as well!) aren't really made with touch screens in mind. There is nothing SteamTinkerLaunch can do about this sadly, we don't draw the UI like Electron apps might. You probably also found that tools like ModOrganizer 2 are somewhat difficult to navigate with a touchscreen/small screen like the Steam Deck - They just weren't made with touchscreens in mind.

The best piece of advice I can give is that you use a mouse or use the right trackpad on Deck to navigate the UI - That is how I personally navigate, instead of using a touch screen. I would also recommend doing tinkering that requires this sort of precision (beyond just checkboxes) in Desktop Mode, or with a mouse in Game Mode.

<hr>

There was an update to change the width of the scrollbar, but it sounds like STL might not be updating to latest Git on Deck anymore based on your other issue? If you go to `$HOME/stl/prefix` and open the `steamtinkerlaunch` file in a text editor like KWrite, and search for `15px`, you'll know if you have the update that's meant to make the scrollbar thinner. Otherwise, if it's `30px`, it might not have applied.

Basically what I want to get at here is: Have you got the change made to make the scrollbar thinner and if so, is it *still* too wide? I don't think making it smaller will help, since then the problem becomes that it's difficult to scroll through the UI (and we can't implement scrolling features - That's up to the windowing toolkit and desktop environment to some extent).

If you do have the thinner scrollbar and it's still too wide, then I'm afraid there's nothing that can be done on STL's side. STL is a graphical tool and graphical tools, no matter Windows, macOS or Linux, were not built with touchscreens in mind. Plus, the Steam Deck uses touch emulation since it runs on X11, so the mouse cursor will always be slightly offset unlike on Wayland which has native touchscreen support.

**Comment by SentaiBrad** [comment]
> There is nothing really that can be done about things being hard to tap, that's kind of the problem with UI apps running on a handheld with a touchscreen. Yad, and by extension regular GTK windowing toolkits (and Qt as well!) aren't really made with touch screens in mind. There is nothing SteamTinkerLaunch can do about this sadly, we don't draw the UI like Electron apps might. You probably also found that tools like ModOrganizer 2 are somewhat difficult to navigate with a touchscreen/small screen like the Steam Deck - They just weren't made with touchscreens in mind.
> 
> The best piece of advice I can give is that you use a mouse or use the right trackpad on Deck to navigate the UI - That is how I personally navigate, instead of using a touch screen. I would also recommend doing tinkering that requires this sort of precision (beyond just checkboxes) in Desktop Mode, or with a mouse in Game Mode.

Agree on the precision aspect. My brain starts looking at all usable aspects. I have a background in IT and Software QA. Currently awaiting my job offer for a CNC Machinist this week :D

As for using Desktop mode: I am not playing games in here for any real length or to enjoy them. It's for testing purposes only. When in the DE, I have a Virtual KVM that I use called [Barrier](https://github.com/debauchee/barrier). Runs on Windows, Linux and Mac, and its awesome. There is a newer fork of this, but I haven't had time to install and test it.
> 
> There was an update to change the width of the scrollbar, but it sounds like STL might not be updating to latest Git on Deck anymore based on your other issue? If you go to `$HOME/stl/prefix` and open the `steamtinkerlaunch` file in a text editor like KWrite, and search for `15px`, you'll know if you have the update that's meant to make the scrollbar thinner. Otherwise, if it's `30px`, it might not have applied.
> 
> Basically what I want to get at here is: Have you got the change made to make the scrollbar thinner and if so, is it _still_ too wide? I don't think making it smaller will help, since then the problem becomes that it's difficult to scroll through the UI (and we can't implement scrolling features - That's up to the windowing toolkit and desktop environment to some extent).
> 
> If you do have the thinner scrollbar and it's still too wide, then I'm afraid there's nothing that can be done on STL's side. STL is a graphical tool and graphical tools, no matter Windows, macOS or Linux, were not built with touchscreens in mind. Plus, the Steam Deck uses touch emulation since it runs on X11, so the mouse cursor will always be slightly offset unlike on Wayland which has native touchscreen support.

Based on the script I have, I am running on 30px. I don't know how I would have gotten *such* an old version running when I've been updating through ProtonUp-Qt. However, yesterday when we had the update notification talk and you made the update, that's when I started getting ProtonUp-Qt stuck at 99%. I assume its related to the issue mentioned on that thread.

In the end, if its at the best its going to get, then so be it. Never do I want to give the air of being ungrateful. It does look like I am running at 30px right now though, so I will wait and see.

**Comment by sonic2kk** [comment]
> I have a background in IT and Software QA. Currently awaiting my job offer for a CNC Machinist this week :D

Awesome, and good luck with your job offer! I appreciate you giving your time in all your issues throughout the last couple of days :-)

> Based on the script I have, I am running on 30px. I don't know how I would have gotten such an old version running when I've been updating through ProtonUp-Qt. However, yesterday when we had the update notification talk and you made the update, that's when I started getting ProtonUp-Qt stuck at 99%. I assume its related to the issue mentioned on that thread.

It could certainly be related to that, I think I may have messed up and force pushed which meant STL can't pull down the latest changes (on Deck it just does a `git pull` with some other fancy steps to set the git dir + `--rebase=false`).

Since you are using the script with `30px`, it would make sense that you think the scrollbar is too wide - I tried to reduce it for this reason. I'll leave this issue open until we can get you on the latest version, then you can give some more feedback on the scrollbar and let me know if you think it's thin enough. *Probably* I could push it down to 12px but that would be at the absolute minimum imo, personally I think 15px is fine but again, I'll leave this open until I can hear your thoughts on that :-)

**Comment by sonic2kk** [comment]
I realise there has been a bunch of context switching, but @SentaiBrad when you get a moment please re-test and see how you feel about the updated scrollbar  that you can (hopefully) use now :-)

**Comment by SentaiBrad** [comment]
It *IS* much better, but still covers UI.

![image](https://user-images.githubusercontent.com/26223818/197604900-c44cc4a9-1473-4a22-b3eb-474e2261fc87.png)

I completely agree though, at a certain point it is too small. Is there a way to shrink the boxes so that they don't take up the entire width? Could they be modified to only be 50-100px long dynamically justified to the left?

**Comment by sonic2kk** [comment]
> Is there a way to shrink the boxes so that they don't take up the entire width?

Oh this is a good idea. I am not sure but there might be. I'm not sure what kinds of GTK properties can be modified with CSS (to be honest the scrollbar code was already set by Frostworx, so I just edited the value to make it thinner).

**Comment by sonic2kk** [comment]
Investigating a bit more, I saw this GTK docs page: https://docs.gtk.org/gtk3/css-overview.html

I had an initial attempt at playing around with the CSS properties here, and the initial results were kind of ugly. I'll keep playing around.

The impression right now is that, editing those UI elements' width *might* be possible, but I am cautious of fiddling too much with their CSS.

**Comment by sonic2kk** [comment]
Okay, I've got some good CSS going here. I still have some more elements I want to modify to keep the UI uniform, but I should have the changes up soon ish. It basically changes the width of the elements so they end a little bit before where the scrollbar starts (adds a `margin-right` of `20px`, scrollbar is `15px`)

**Comment by sonic2kk** [comment]
Okay, I pushed up a change that should help with some of the UI elements being covered by the scrollbar :+1:

**Comment by sonic2kk** [comment]
@SentaiBrad have you had a chance to revisit this at all?

**Comment by SentaiBrad** [comment]
> @SentaiBrad have you had a chance to revisit this at all?

 Apologies, real life and all. Looking at it now, I see that the boxes for input in the Main Menu Config look different, but they are still underneath the scroll bar. If this is the extent to which it can be controlled, so be it. I did label this a Small UI issue xD. It overall looks a lot better than it was.

![image](https://user-images.githubusercontent.com/26223818/198714227-10aa79c6-dfac-4b33-98fb-7128bb79e896.png)

**Comment by sonic2kk** [comment]
That screenshot is not showing the UI updates it looks like. It should be a lot smaller. Are you sure you have updated to the latest SteamTinkerLaunch?

![image](https://user-images.githubusercontent.com/7917345/198715430-0a689cda-dabb-4f07-918e-5d2f4f407515.png)

**Comment by SentaiBrad** [comment]
> That screenshot is not showing the UI updates it looks like. It should be a lot smaller. Are you sure you have updated to the latest SteamTinkerLaunch?

Dumb. I re-applied ProtonUp-Qt for the master branch and I am 100% up-to-date now, Looks a hell of a lot better. 
![image](https://user-images.githubusercontent.com/26223818/198715682-a5dc149b-6d64-49f2-9fcc-43b174745387.png)

**Comment by sonic2kk** [comment]
No worries, easy mistake :-) 

Glad to know you think it's an improvement, and it even looks fine with your custom theme! Very relieved about that :sweat_smile: 

Concerning the changes, I could make the scrollbar wider and add some more padding to the sides, but I wanted to make the scrollbar slightly thinner anyway. I think right now the UI has a nice balance, but my main concern is that it's not very "Uniform". The text entry boxes and file browser buttons (like "custom command" and "custom command arguments") do not have the margin that the other fields have. This is because I couldn't find the selector for these elements to set in the CSS.

I have plans to change how SteamTinkerLaunch exports this. I would like to store the CSS in a file and move that file instead of echoing out the contents to `gtk-3.0.css`. This would also mean users could set their own custom CSS profiles and load them with SteamTinkerLaunch, even on desktop. It would also make contributions to improving the UI on Deck much more straightforward, and we could even include UI profiles (such as one with a wider scrollbar and more padding!)

There are a few other concerns around this and it's not a very high priority, but the UI I think now is "good enough" to be considered an improvement. I can say personally it's been nicer to use the UI on Deck after you brought these changes up, it pushed me to get off my ass and do something :wink: 

So while improvements could be made, I think for now this issue is "resolved." I will revisit the UI hopefully at some point but I will leave it up to you whether you feel this is resolved and ready to be closed. Please feel free to give more feedback, but if you're happy with the current UI changes we can close this issue :slightly_smiling_face:
